resolver/dns/dns_resolver_go19_test.go

```diff
@@ -0,0 +1,133 @@
+// +build go1.9
+
+/*
+ *
+ * Copyright 2018 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package dns
+
+import (
+	"errors"
+	"net"
+	"testing"
+
+	"golang.org/x/net/context"
+	"google.golang.org/grpc/internal/leakcheck"
+	"google.golang.org/grpc/resolver"
+)
+
+func TestCustomAuthority(t *testing.T) {
+	defer leakcheck.Check(t)
+
+	tests := []struct {
+		authority     string
+		authorityWant string
+		expectError   bool
+	}{
+		{
+			"4.3.2.1:53",
+			"4.3.2.1:53",
+			false,
+		},
+		{
+			"4.3.2.1:123",
+			"4.3.2.1:123",
+			false,
+		},
+		{
+			"4.3.2.1",
+			"4.3.2.1:53",
+			false,
+		},
+		{
+			"::1",
+			"[::1]:53",
+			false,
+		},
+		{
+			"[::1]",
+			"[::1]:53",
+			false,
+		},
+		{
+			"[::1]:123",
+			"[::1]:123",
+			false,
+		},
+		{
+			"dnsserver.com",
+			"dnsserver.com:53",
+			false,
+		},
+		{
+			":123",
+			"localhost:123",
+			false,
+		},
+		{
+			":",
+			"",
+			true,
+		},
+		{
+			"[::1]:",
+			"",
+			true,
+		},
+		{
+			"dnsserver.com:",
+			"",
+			true,
+		},
+	}
+	oldCustomAuthorityDialler := customAuthorityDialler
+	defer func() {
+		customAuthorityDialler = oldCustomAuthorityDialler
+	}()
+
+	for _, a := range tests {
+
+		customAuthorityDialler = func(authority string) func(ctx context.Context, network, address string) (net.Conn, error) {
+
+			return func(ctx context.Context, network, address string) (net.Conn, error) {
+				if authority != a.authorityWant {
+					t.Errorf("wrong custom authority passed to resolver. input: %s expected: %s actual: %s", a.authority, a.authorityWant, authority)
```

Comment:
Since the the dialler will be called in a separate goroutine(i.e. the watcher goroutine), the `Errorf` here may not be called before the main test goroutine exit.

The better solution here is to create a channel of `error` type. And inside the custom dial function, push an error(or nil for no error) to this channel. And in the main goroutine, wait until an error is pushed to the channel, and check whether it is nil or not.
```go
for _, a := range tests {
	errChan := make(chan error, 1)
	customAuthorityDialler = func(...) {
		...
    	return func(ctx context.Context, network, address string) (net.Conn, error) {
	  		if authority != a.authorityWant {
        		errChan <- fmt.Errorf(...)
			} else {
				errChan <- nil
			}
		...
	}
    ...
	// wait until the check in the dial function completes
	err := <- errChan
	// verify whether there's an error or not.
	if err != nil {
		t.Errorf(...)
	}
}              
```